### PR TITLE
Add support for data bound symbols as attribute names.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Took the HTML5 spec approach of disallowing specific characters and accepting all others.
- Added several parser and code generation tests to cover both `TagHelper` and non-`TagHelper` variations of symbol bound attribute names.

#137

### CODE REVIEW

File: src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs [hunk 1]
```diff
@@ -1128,5 +1128,24 @@ private void EndTagBlock(Stack<Tuple<HtmlSymbol, SourceLocation>> tags, bool com
...
         }
+
+        internal static bool IsValidAttributeNameSymbol(HtmlSymbol symbol)
+        {
+            if (symbol == null)
+            {
+                return false;
+            }
+
+            var symbolType = symbol.Type;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs]: Long list and perhaps the `... == ... || ...` list would be (a bit) longer. Nonetheless suggest switching to the positive list for those of us who haven't memorized `HtmlSymbolType`.

**Comment by ajaybhargavb** [comment]
Maybe add one or two tests with newlines and whitespace before and after the attribute name.

**Comment by ajaybhargavb** [comment]
:watch:

File: src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs]: Nah, spec defines it via the negative.

**Comment by NTaylorMullen** [comment]
Updated.

File: test/Microsoft.AspNet.Razor.Test/Parser/Html/HtmlAttributeTest.cs [hunk 2]
```diff
@@ -14,6 +14,112 @@ namespace Microsoft.AspNet.Razor.Test.Parser.Html
...
+                lineIndex: 2,
+                characterIndex: 5 + attributeNameLength);
+            var valueLocation2 = new SourceLocation(
+                absoluteIndex: 14 + attributeNameLength * 2 + newlineLength * 2,
+                lineIndex: 2,
+                characterIndex: 2 + attributeNameLength);
+
+            // Act & Assert
+            ParseBlockTest(
+                $"<a {Environment.NewLine}  {attributeName}='Foo'\t{Environment.NewLine}{attributeName}='Bar' />",
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Razor.Test/Parser/Html/HtmlAttributeTest.cs]: Suggest adding one more variation with Environment.NewLine before =.

**Comment by ajaybhargavb** [comment]
:shipit:

File: src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs [hunk 3]
```diff
@@ -1136,5 +1136,24 @@ private void EndTagBlock(Stack<Tuple<HtmlSymbol, SourceLocation>> tags, bool com
...
+            return symbolType != HtmlSymbolType.WhiteSpace &&
+                symbolType != HtmlSymbolType.NewLine &&
+                symbolType != HtmlSymbolType.CloseAngle &&
+                symbolType != HtmlSymbolType.OpenAngle &&
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs]: `OpenAngle` is legal according to http://www.w3.org/TR/html5/syntax.html#attributes-0

File: src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs]: :ok: how 'bout a link to http://www.w3.org/TR/html5/syntax.html#attributes-0 ?

File: src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs [hunk 4]
```diff
@@ -1136,5 +1136,24 @@ private void EndTagBlock(Stack<Tuple<HtmlSymbol, SourceLocation>> tags, bool com
...
+                symbolType != HtmlSymbolType.ForwardSlash &&
+                symbolType != HtmlSymbolType.DoubleQuote &&
+                symbolType != HtmlSymbolType.SingleQuote &&
+                symbolType != HtmlSymbolType.Equals &&
+                symbolType != HtmlSymbolType.Unknown;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs]: `HtmlSymbolType` doesn't line up with HTML character classification. This expression allows `HtmlSymbolType.Transition`, `RazorComment`, `RazorCommentStar`, and `RazorCommentTransition`. None of those seem correct for attribute names in the Razor context.

Further `HtmlSymbolType.Text` (currently allowed) indicates _any_ non-whitespace non-symbol series of characters. Need to exclude control characters and unknown Unicode characters when that's the `symbolType`.

File: test/Microsoft.AspNet.Razor.Test/Parser/Html/HtmlAttributeTest.cs [hunk 5]
```diff
@@ -14,6 +14,112 @@ namespace Microsoft.AspNet.Razor.Test.Parser.Html
...
+            get
+            {
+                return new TheoryData<string>
+                {
+                    "[item]",
+                    "[(item,",
+                    "(click)",
+                    "(^click)",
+                    "*something",
+                    "#local",
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/Parser/Html/HtmlAttributeTest.cs]: Need to test more than just symbols Angular would allow. Non-Latin1 characters at least.

I also don't see tests verifying rejection of anything but whitespace and `=`.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs]: WIll add

File: src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs (hunk 3 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs]: Yup, Razor don't like it though :smile:. This is an "almost" spec worthy fix. 

This PR solves all of the use-cases brought forward in issue #137 and takes an extra step in making it spec equivalent. I just filed an issue to cover the open angle derivation: https://github.com/aspnet/Razor/issues/538

File: src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs (hunk 4 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs]: Talked in person.

File: test/Microsoft.AspNet.Razor.Test/Parser/Html/HtmlAttributeTest.cs (hunk 5 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/Parser/Html/HtmlAttributeTest.cs]: Talked in person.

**Comment by NTaylorMullen** [comment]
Updated

**Comment by dougbu** [comment]
Would still like to see a couple of negative tests. Show attribute name parsing stopping at excluded characters. E.g. confirm `name>` or `name'` are parsed as the attribute name `name` followed by a separate (garbage) symbol. Otherwise this PR LGTM.

**Comment by dougbu** [comment]
Talked offline. The tests I wanted already exist and didn't need to be changed in this PR.

**Comment by dougbu** [comment]
:shipit:
